### PR TITLE
[MNG-7629] Change reactor reader to copy packaged artifacts and reuse them across builds if needed

### PR DESCRIPTION
**Pull request by gnodet**
Similar to #912 but without removing the best effort artifact resolution to keep `mvn test` working _if possible_.

This require a few ITs to be modified:
  https://github.com/apache/maven-integration-testing/pull/234/

### CODE REVIEW

**Review by mthmulders — COMMENTED** [review]
Left a couple of questions.

Also, I wonder if this change allows us to remove the fallbacks that were previously on line 150 - 159. Since those lines aren't changed, it seems GitHub doesn't allow me to comment on the spot :-\\.

File: maven-core/src/main/java/org/apache/maven/ReactorReader.java
Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/ReactorReader.java:227]: The idea is to run at least the `package` phase, but I thought we wanted to promote running `mvn verify` rather than `mvn install`.

File: maven-core/src/main/java/org/apache/maven/ReactorReader.java
Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/ReactorReader.java:351]: You're right, it cannot.

File: maven-core/src/main/java/org/apache/maven/ReactorReader.java
Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/ReactorReader.java:350]: I don't think so.  The copying of artifacts should not really alter the current build, as those artifacts are already packaged anyway and the original packaged artifacts should be used.  It would mainly affect a subsequent build, which may fail as it can do now, so I'd think a warning is sufficient.

File: maven-core/src/main/java/org/apache/maven/ReactorReader.java
Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/ReactorReader.java]: Raised to info level.

File: maven-core/src/main/java/org/apache/maven/ReactorReader.java
Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/ReactorReader.java:389]: Fixed in the javadoc description for the `getProjects()` method and on the `projects` field.

File: maven-core/src/main/java/org/apache/maven/ReactorReader.java
Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/ReactorReader.java]: Fixed

File: maven-core/src/main/java/org/apache/maven/ReactorReader.java
Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/ReactorReader.java]: Renamed

File: maven-core/src/main/java/org/apache/maven/ReactorReader.java
Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/ReactorReader.java]: > Also, I wonder if this change allows us to remove the fallbacks that were previously on line 150 - 159. Since those lines aren't changed, it seems GitHub doesn't allow me to comment on the spot :-\.

Good question.  I think part of it could be removed, we'll need to keep the `determineBuildOutputDirectoryForArtifact` in order to be able to partially run `mvn test`.

**Comment by gnodet** [comment]
> Left a couple of questions.
> 
> Also, I wonder if this change allows us to remove the fallbacks that were previously on line 150 - 159. Since those lines aren't changed, it seems GitHub doesn't allow me to comment on the spot :-\.

Good question. I think part of it could be removed, but we need to keep the `determineBuildOutputDirectoryForArtifact` in order to be able to partially run mvn test.

Another possible strategy instead of copying the artifacts would be to only record their paths.

File: maven-core/src/main/java/org/apache/maven/ReactorReader.java
Comment (mthmulders) [reply maven-core/src/main/java/org/apache/maven/ReactorReader.java:392]: Thanks!

**Comment by mthmulders** [comment]
> ... in order to be able to partially run mvn test.

I'm not sure I understand. You mean "run `mvn test` on a multi-module project without requiring the user to have ran `mvn package` first"?

**Comment by gnodet** [comment]
> > ... in order to be able to partially run mvn test.
> 
> I'm not sure I understand. You mean "run `mvn test` on a multi-module project without requiring the user to have ran `mvn package` first"?

Yes.  I haven't removed this part in this PR.  If the main artifacts have not been packaged, they will point to the `target/classes`.

File: maven-core/src/main/java/org/apache/maven/ReactorReader.java
Comment (MartinKanters) [inline maven-core/src/main/java/org/apache/maven/ReactorReader.java:341]: You could go for streams here as well, if you like.

**Review by MartinKanters — COMMENTED** [review]
I like the solution in general. Added small comments.

File: maven-core/src/main/java/org/apache/maven/ReactorReader.java
Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/ReactorReader.java]: I've renamed a few methods.

**Comment by gnodet** [comment]
> > > ... in order to be able to partially run mvn test.
> > 
> > 
> > I'm not sure I understand. You mean "run `mvn test` on a multi-module project without requiring the user to have ran `mvn package` first"?
> 
> Yes. I haven't removed this part in this PR. If the main artifacts have not been packaged, they will point to the `target/classes`.

I've been able to remove [the check to load the packaged artifacts](https://github.com/apache/maven/pull/954/commits/197b743db736adf419627f56456e2d4b13f2d2e7).

**Comment by gnodet** [comment]
In the internal project local repository structure, I've added a child directory with the version of a given artifact.  This gives a structure which is closer to the maven local repository, and it allows deleting artifacts during the clean phase without having to rely on a heuristic to split between classifier / version / extension.

@mthmulders @MartinKanters I'm done on this PR, so unless you have further comments, an approval before the merge would be nice !

**Review by cstamas — APPROVED** [review]

File: maven-core/src/main/java/org/apache/maven/ReactorReader.java
Comment (cstamas) [inline maven-core/src/main/java/org/apache/maven/ReactorReader.java:374]: um, this is hardcoded?

File: maven-core/src/main/java/org/apache/maven/ReactorReader.java
Comment (cstamas) [inline maven-core/src/main/java/org/apache/maven/ReactorReader.java:373]: Also, the use of `session.getRequest().getMultiModuleProjectDirectory()` implies existence of top level `.mvn`?

File: maven-core/src/main/java/org/apache/maven/ReactorReader.java
Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/ReactorReader.java:374]: I'm not really sure this is configurable.  Is it ?

Comment (cstamas) [reply maven-core/src/main/java/org/apache/maven/ReactorReader.java:374]: https://maven.apache.org/pom.html#Build

Comment (cstamas) [reply maven-core/src/main/java/org/apache/maven/ReactorReader.java:374]: Default is in super POM https://github.com/apache/maven/blob/20f7c65a520ee4c6a5dde0de4d58e304b1bde442/maven-model-builder/src/main/resources/org/apache/maven/model/pom-4.0.0.xml#L54

File: maven-core/src/main/java/org/apache/maven/ReactorReader.java
Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/ReactorReader.java:373]: First, [this property is always available](https://github.com/apache/maven/blob/c2c6dd09219744cfe45df34cabed6684fe5c9213/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java#L308-L320), and second, if no `.mvn` parent directory exists, the current directory will be used.  So this is maybe not perfect, and I'd really like to have this root/top directory property problem solved in the near future, but in the mean time, that's the best we have afaik.

Comment (cstamas) [reply maven-core/src/main/java/org/apache/maven/ReactorReader.java:373]: Cool, was unaware it is always available... and big :+1:  to solve it in generic way

File: maven-core/src/main/java/org/apache/maven/ReactorReader.java
Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/ReactorReader.java:374]: I could use something like the following to retrieve the build directory of the top most project if it's in the reactor:
```
    private Path getProjectLocalRepo() {
        File mmpd = session.getRequest().getMultiModuleProjectDirectory();
        return session.getProjects().stream()
                .filter(project -> mmpd.equals(project.getBasedir()))
                .findFirst()
                .map(project -> project.getBuild().getDirectory())
                .map(Paths::get)
                .orElseGet(() -> mmpd.toPath().resolve("target"))
                .resolve("project-local-repo");
    }
```

Comment (gnodet) [reply maven-core/src/main/java/org/apache/maven/ReactorReader.java]: Fixed

**Review by MartinKanters — APPROVED** [review]
Looks like an elegant solution to me. Nice job.

**Review by cstamas — APPROVED** [review]

**Review by mthmulders — APPROVED** [review]
> I've been able to remove [the check to load the packaged artifacts](https://github.com/apache/maven/pull/954/commits/197b743db736adf419627f56456e2d4b13f2d2e7).

Awesome.

> In the internal project local repository structure, I've added a child directory with the version of a given artifact. This gives a structure which is closer to the maven local repository, and it allows deleting artifacts during the clean phase without having to rely on a heuristic to split between classifier / version / extension.

Also awesome.

Happy to see this improvement. Thanks a lot @gnodet!

**Comment by jira-importer** [comment]
Resolve #8824

**Pull request merged**
